### PR TITLE
Fix nll_loss to support empty tensors on GPU

### LINKED ISSUE #31472: nll_loss doesn't support empty tensors on gpu
## 🐛 Bug

Passing empty tensors to nll_loss should return an empty tensor. It is the case on cpu, but crashes on gpu

## To Reproduce

```
>>> import torch
>>> a = torch.rand(0, 4).cuda()
>>> b = torch.tensor([]).cuda()
>>> torch.nn.functional.nll_loss(a, b.long(), reduction='none')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/private/home/alcinos/.conda/envs/detection/lib/python3.6/site-packages/torch/nn/functional.py", line 1845, in nll_loss
    ret = torch._C._nn.nll_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index)
RuntimeError: invalid argument 2: non-empty vector or matrix expected at /opt/conda/conda-bld/pytorch_1576310878804/work/aten/src/THCUNN/generic/ClassNLLCriterion.cu:31
```

## Expected behavior

On cpu it works well:
```
>>> a = torch.rand(0, 4)
>>> b = torch.tensor([])
>>> torch.nn.functional.nll_loss(a, b.long(), reduction='none')
tensor([])
```

## Environment

PyTorch version: 1.4.0.dev20191214
Is debug build: No
CUDA used to build PyTorch: 9.2

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.2.88
GPU models and configuration:
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 410.79
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.17.3
[pip] pytorch-memlab==0.0.4
[pip] torch==1.4.0.dev20191214
[pip] torchfile==0.1.0
[pip] torchnet==0.0.4
[pip] torchvision==0.5.0a0+4f2e755
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      243
[conda] mkl-service               2.3.0            py36he904b0f_0
[conda] mkl_fft                   1.0.15           py36ha843d7b_0
[conda] mkl_random                1.1.0            py36hd6b4f25_0
[conda] pytorch                   1.4.0.dev20191214 py3.6_cuda9.2.148_cudnn7.6.3_0    pytorch-nightly
[conda] pytorch-memlab            0.0.4                    pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchnet                  0.0.4                    pypi_0    pypi
[conda] torchvision               0.5.0.dev20191214       py36_cu92    pytorch-nightly

cc @ngimel

### PR DESCRIPTION
**Pull request by pbelevich**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31491 Fix nll_loss to support empty tensors on GPU**

Fixes #31472

Differential Revision: [D19537231](https://our.internmc.facebook.com/intern/diff/D19537231)

### CODE REVIEW

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Please add tests for cpu and cuda

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 9934633:


* **2/2** failures introduced in this PR
* **2/2** recognized as flaky
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=9934633361b51dfb79adb3879791f283b912933a).

### :detective: 2 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4020331) pytorch_windows_test2 (1/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=77678969) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=77678969))
<details>
<summary>
<code> - cmake (exited 1) - cmake not installed. An error occurred during installation:</code>
</summary>

```
 The remote server returned an error: (503) Server Unavailable. Service Unavailable 
cmake package files install completed. Performing other installation steps. 
The install of cmake was NOT successful. 
cmake not installed. An error occurred during installation: 
 The remote server returned an error: (503) Server Unavailable. Service Unavailable 
 
Chocolatey installed 0/1 packages. 1 packages failed. 
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log). 
 
Failures 
 - cmake (exited 1) - cmake not installed. An error occurred during installation: 
 The remote server returned an error: (503) Server Unavailable. Service Unavailable 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4020332) pytorch_windows_test1 (2/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=77679976) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=77679976))
<details>
<summary>
<code> - cmake (exited 1) - cmake not installed. An error occurred during installation:</code>
</summary>

```
 The remote server returned an error: (503) Server Unavailable. Service Unavailable 
cmake package files install completed. Performing other installation steps. 
The install of cmake was NOT successful. 
cmake not installed. An error occurred during installation: 
 The remote server returned an error: (503) Server Unavailable. Service Unavailable 
 
Chocolatey installed 0/1 packages. 1 packages failed. 
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log). 
 
Failures 
 - cmake (exited 1) - cmake not installed. An error occurred during installation: 
 The remote server returned an error: (503) Server Unavailable. Service Unavailable 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: aten/src/THCUNN/generic/ClassNLLCriterion.cu [hunk 1]
```diff
@@ -28,7 +28,9 @@ void THNN_(ClassNLLCriterion_updateOutput)(
     );
   }
 
-  THArgCheck(!input->is_empty() && (n_dims <= 2 && n_dims > 0), 2, "non-empty vector or matrix expected");
+  if (n_dims != 1 && n_dims != 2) {
+    THError("input tensor should be 1D or 2D");
```
Comment (gchanan) [inline aten/src/THCUNN/generic/ClassNLLCriterion.cu]: doesn't the backwards also need to be updated?  (add a test).

File: aten/src/THCUNN/generic/ClassNLLCriterion.cu [hunk 2]
```diff
@@ -44,24 +46,25 @@ void THNN_(ClassNLLCriterion_updateOutput)(
...
-        toDeviceTensor<scalar_t, 1>(state, output),
-        weights ? THCTensor_(data)(state, weights) : NULL,
-        n_classes,
-        ignore_index);
-
-    THCudaCheck(cudaGetLastError());
-
-    if (weights) {
-      THCTensor_(free)(state, weights);
+    if (batch_size != 0) {
```
Comment (zou3519) [inline aten/src/THCUNN/generic/ClassNLLCriterion.cu]: nit: I think this is clearer as
```
if (batch_size == 0) {
    return;
}
```
An added advantage of this approach is that we don't end up modifying the blame for the rest of the code.

There's are a few other places in the code that would benefit from this idiom.

Comment (zou3519) [inline aten/src/THCUNN/generic/ClassNLLCriterion.cu]: We should add a comment that this is to workaround the CUDA launch parameters not handling 0

File: test/test_nn.py [hunk 3]
```diff
@@ -10414,6 +10401,47 @@ def test_multi_margin_loss_errors(self, device):
...
+        C = 10
+        for N in [0, 10]:
+            x = torch.rand(N, C, requires_grad=True, device=device)
+            target = torch.randint(C, (N,), device=device)
+            for reduction in ['none', 'mean', 'sum']:
+                y = F.nll_loss(x, target, reduction=reduction)
+                if reduction == 'none':
+                    self.assertTrue(y.size() == target.size())
+                else:
+                    self.assertEqual(y.dim(), 0)
```
Comment (zou3519) [inline test/test_nn.py]: If reduction is not "none", then we should check `self.assertEqual(y.item(), 0)` as well (to make sure the value is correct).

File: test/test_nn.py [hunk 4]
```diff
@@ -10414,6 +10401,47 @@ def test_multi_margin_loss_errors(self, device):
...
+                y.sum().backward()
```
Comment (zou3519) [inline test/test_nn.py]: We should check that `y.grad.size() == y.size()` to complete our check that the backwards pass handles the 0-size batch dim case

File: test/test_nn.py [hunk 5]
```diff
@@ -10414,6 +10401,47 @@ def test_multi_margin_loss_errors(self, device):
...
+
+    def test_nll_loss_2d(self, device):
+        C = 10
+        H = 10
+        W = 10
+        for N in [0, 10]:
+            x = torch.rand(N, C, H, W, requires_grad=True, device=device)
+            target = torch.randint(C, (N, H, W), device=device)
+            for reduction in ['none', 'mean', 'sum']:
```
Comment (zou3519) [inline test/test_nn.py]: Question: The code here is almost identical to `test_nll_loss` aside from the creation of `x` and `target`, would it be nicer to merge `test_nll_loss2d` with `test_nll_loss` ?

File: test/test_nn.py [hunk 6]
```diff
@@ -10414,6 +10401,47 @@ def test_multi_margin_loss_errors(self, device):
...
+                y = F.nll_loss(x, target, reduction=reduction)
+                if reduction == 'none':
+                    self.assertTrue(y.size() == target.size())
+                else:
```
Comment (zou3519) [inline test/test_nn.py]: The commit message and PR title says: "Fix nll_loss to support empty tensors on GPU". I'm not sure that nll_loss supports all empty tensors; this PR fixes nll_loss to support tensors with size-zero batch. We should either change the commit message or test that nll_loss supports all empty tensors (tensors with height or width or channels = 0).

**Review by zou3519 — CHANGES_REQUESTED** [review]
Code looks fine, we should improve the testing

File: aten/src/THCUNN/generic/SpatialClassNLLCriterion.cu [hunk 7]
```diff
@@ -33,8 +33,8 @@ static void THNN_(SpatialClassNLLCriterion_gradOutput_no_reduce_shapeCheck)(
            THCTensor *gradOutput,
            THCIndexTensor *target)
 {
-  TORCH_CHECK(!gradOutput->is_empty() && THCTensor_(nDimensionLegacyNoScalars)(state, gradOutput) == 3, 2,
-           "Expected non-empty dimension 3 but got gradOutput of size: ", gradOutput->sizes());
+  TORCH_CHECK(THCTensor_(nDimensionLegacyNoScalars)(state, gradOutput) == 3, 2,
+    "grad_output must have same dimension as target (3) but got dimension: ", gradOutput->sizes());
```
[the remark this answers is not in the document]
Comment (pbelevich) [reply aten/src/THCUNN/generic/SpatialClassNLLCriterion.cu]: addressed

File: aten/src/THCUNN/generic/ClassNLLCriterion.cu (hunk 2 above)
Comment (pbelevich) [reply aten/src/THCUNN/generic/ClassNLLCriterion.cu]: addressed

Comment (pbelevich) [reply aten/src/THCUNN/generic/ClassNLLCriterion.cu]: addressed

File: test/test_nn.py (hunk 3 above)
Comment (pbelevich) [reply test/test_nn.py]: addresed

File: test/test_nn.py (hunk 4 above)
Comment (pbelevich) [reply test/test_nn.py]: addressed

File: test/test_nn.py (hunk 5 above)
Comment (pbelevich) [reply test/test_nn.py]: addressed

File: test/test_nn.py [hunk 8]
```diff
@@ -10414,6 +10401,47 @@ def test_multi_margin_loss_errors(self, device):
...
+        with self.assertRaisesRegex(ValueError, 'Expected.*batch_size'):
+            F.nll_loss(x, t)
+
+    def test_nll_loss_out_of_bounds_ignore_index(self, device):
+        x = torch.randn(6, 3, requires_grad=True, device=device)
+        t = torch.tensor([0, 1, 255, 0, 1, 2], dtype=torch.int64, device=device)
+        for reduction in ['mean', 'none']:
+            F.nll_loss(x, t, ignore_index=255, reduction=reduction).sum().backward()
+
+    def test_nll_loss(self, device):
```
[the remark this answers is not in the document]
Comment (pbelevich) [reply test/test_nn.py]: changed to empty tensors only

File: test/test_nn.py (hunk 6 above)
Comment (pbelevich) [reply test/test_nn.py]: I covered cases when H = 0 and W =0. nll_loss does not make sense if C = 0 afaik

**Review by zou3519 — CHANGES_REQUESTED** [review]

File: aten/src/THCUNN/generic/SpatialClassNLLCriterion.cu [hunk 9]
```diff
@@ -169,23 +175,25 @@ void THNN_(SpatialClassNLLCriterion_updateGradInput)(
...
-        count,
-        toDeviceTensor<THCIndex_t, 3>(state, target),
-        toDeviceTensor<scalar_t, 3>(state, gradOutput),
-        toDeviceTensor<scalar_t, 4>(state, gradInput),
-        weights ? THCTensor_(data)(state, weights) : NULL,
-        ignore_index);
 
-    if (weights) {
-      THCTensor_(free)(state, weights);
+    if (count != 0) {
```
[the remark this answers is not in the document]
Comment (pbelevich) [reply aten/src/THCUNN/generic/SpatialClassNLLCriterion.cu]: addressed

File: test/test_nn.py [hunk 10]
```diff
@@ -10426,6 +10413,54 @@ def test_multi_margin_loss_errors(self, device):
...
+
+        N = 0
+        x = torch.rand(N, C, requires_grad=True, device=device)
+        target = torch.randint(C, (N,), device=device)
+        test_nll_loss(x, target)
+
+        # 2-D scenarios
+        for N in [0, 10]:
+            for H in [0, 10]:
+                for W in [0, 10]:
```
[the remark this answers is not in the document]
Comment (pbelevich) [reply test/test_nn.py]: This test is running NLLLoss twice

File: test/test_nn.py [hunk 11]
```diff
@@ -10426,6 +10413,54 @@ def test_multi_margin_loss_errors(self, device):
...
+
+        C = 10
+
+        N = 0
+        x = torch.rand(N, C, requires_grad=True, device=device)
+        target = torch.randint(C, (N,), device=device)
+        test_nll_loss(x, target)
+
+        # 2-D scenarios
+        for N in [0, 10]:
```
[the remark this answers is not in the document]
Comment (zou3519) [reply test/test_nn.py]: Er nevermind; I realized we're also trying to test with N = 10 and, say, W = 0.

File: test/test_nn.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply test/test_nn.py]: Oh you're completely right, that is my bad

File: test/test_nn.py [hunk 12]
```diff
@@ -10426,6 +10413,54 @@ def test_multi_margin_loss_errors(self, device):
...
+            for H in [0, 10]:
+                for W in [0, 10]:
+                    if (N * C * H * W == 0): # If input is empty
+                        x = torch.rand(N, C, H, W, requires_grad=True, device=device)
+                        if C != 0:
+                            target = torch.randint(C, (N, H, W), device=device)
+                        else:
+                            target = torch.empty((N, H, W), dtype=torch.int64, device=device)
```
[the remark this answers is not in the document]
Comment (pbelevich) [reply test/test_nn.py]: addressed

File: aten/src/THCUNN/generic/SpatialClassNLLCriterion.cu [hunk 13]
```diff
@@ -102,34 +106,36 @@ void THNN_(SpatialClassNLLCriterion_updateOutput)(
...
-      THCTensor_(size)(state, input, 1),
-      THCTensor_(size)(state, input, 2) * THCTensor_(size)(state, input, 3),
-      blocks_per_sample,
-      ignore_index
-  );
-  THCudaCheck(cudaGetLastError());
-  if (reduction == at::Reduction::Mean) {
-    cunn_SpatialClassNLLCriterion_sizeAverage_kernel<<<1, 1, 0, THCState_getCurrentStream(state)>>>(
-      output_data, total_weight_data
+  if (batch_size != 0) { // This guards from unnecessary operations and launching CUDA kernel with 0 blocks.
```
[the remark this answers is not in the document]
Comment (pbelevich) [reply aten/src/THCUNN/generic/SpatialClassNLLCriterion.cu]: addressed

File: test/test_nn.py [hunk 14]
```diff
@@ -10426,6 +10413,54 @@ def test_multi_margin_loss_errors(self, device):
...
+
+    def test_nll_loss_out_of_bounds_ignore_index(self, device):
+        x = torch.randn(6, 3, requires_grad=True, device=device)
+        t = torch.tensor([0, 1, 255, 0, 1, 2], dtype=torch.int64, device=device)
+        for reduction in ['mean', 'none']:
+            F.nll_loss(x, t, ignore_index=255, reduction=reduction).sum().backward()
+
+    def test_nll_loss_empty_tensor(self, device):
+
+        def test_nll_loss(x, target):
```
[the remark this answers is not in the document]
Comment (pbelevich) [reply test/test_nn.py]: agree, I changed the code

File: aten/src/THCUNN/SpatialClassNLLCriterion.cu [hunk 15]
```diff
@@ -123,8 +123,9 @@ __global__ void cunn_SpatialClassNLLCriterion_sizeAverage_kernel(
           T *output,
           T *total_weight)
 {
-  if (*total_weight > 0)
+  if (*total_weight >= 0) {
```
Comment (zou3519) [inline aten/src/THCUNN/SpatialClassNLLCriterion.cu]: Yeah, I feel like the original condition, `*total_weight > 0`, was used to guard against division by zero and we might not have to worry about negative numbers.

I'm a little worried about BC -- I'm not sure if any users relied on the total_weight = 0 behavior

**Review by zou3519 — COMMENTED** [review]
Code and tests look good. I'm worried about if users might have relied on the total_weight = 0 behavior and if we're breaking anyone with this

File: aten/src/THCUNN/ClassNLLCriterion.cu [hunk 16]
```diff
@@ -125,7 +125,7 @@ __global__ void cunn_ClassNLLCriterion_updateOutput_kernel(Dtype *output,
     }
     *total_weight = ScalarConvert<Acctype, Dtype>::to(total_weightAcc);
     *output = ScalarConvert<Acctype, Dtype>::to(outputAcc);
-    if (size_average && *total_weight > 0) {
+    if (size_average && *total_weight >= 0) {
```
Comment (zou3519) [inline aten/src/THCUNN/ClassNLLCriterion.cu]: There's some code in  cunn_ClassNLLCriterion_updateOutput_kernel1 that does the same thing, do we need to update that? https://github.com/pytorch/pytorch/blob/638e4ad8b95793876c0f5e82db58c18618fb6e61/aten/src/THCUNN/ClassNLLCriterion.cu#L35-L37

File: aten/src/THCUNN/SpatialClassNLLCriterion.cu (hunk 15 above)
Comment (zou3519) [reply aten/src/THCUNN/SpatialClassNLLCriterion.cu]: So, concrete action items:
(1) we should try to find some example of this type of reduction, dividing by weights, in another framework and see what happens
(2) we should think about a use case where total_weight = 0 and use that to inform our decision of if we want to return 0 or NaN

Comment (zou3519) [reply aten/src/THCUNN/SpatialClassNLLCriterion.cu]: I investigated this. Let me rephrase my findings.

File: aten/src/THCUNN/ClassNLLCriterion.cu (hunk 16 above)
Comment (zou3519) [inline aten/src/THCUNN/ClassNLLCriterion.cu]: Before this PR, the following test case passes:
```
import torch
import torch.nn.functional as F

input = torch.ones(2, 3)
target = torch.zeros(2, dtype=torch.long)
weight = torch.zeros(3)
assert F.nll_loss(input, target, weight).item() == 0
assert F.nll_loss(input.cuda(), target.cuda(), weight.cuda()).item() == 0
```

After this PR, the above test case fails. In particular, `F.nll_loss(input.cuda(), target.cuda(), weight.cuda())` is now equal to NaN.

This is a pretty weird edge case (that we don't have a test for), but we should not change the behavior. Let's (1) add a test case for this somewhere and (2) fix it

**Review by zou3519 — CHANGES_REQUESTED** [review]

File: test/test_nn.py [hunk 17]
```diff
@@ -10431,6 +10418,50 @@ def test_multi_margin_loss_errors(self, device):
...
+
+            output.sum().backward()
+            self.assertEqual(input.grad.size(), input.size())
+
+        helper([0, 3], "none", torch.empty([0], device=device))
+        helper([0, 3, 5, 7], "none", torch.empty([0, 5, 7], device=device))
+        helper([2, 3, 0, 7], "none", torch.empty([2, 0, 7], device=device))
+        helper([2, 3, 5, 0], "none", torch.empty([2, 5, 0], device=device))
+
+        nan = torch.tensor([float('nan')], device=device).resize_([])
```
[the remark this answers is not in the document]
Comment (pbelevich) [reply test/test_nn.py]: Correct me if I'm wrong, but `torch.tensor([float('nan')]) = tensor([nan])` and size is [1]. If I resize it then it became `tensor(nan)` with size []. I expect size to be [], not [1].

File: test/test_nn.py [hunk 18]
```diff
@@ -10431,6 +10418,50 @@ def test_multi_margin_loss_errors(self, device):
...
+        helper([0, 3], "mean", nan)
+        helper([0, 3, 5, 7], "mean", nan)
+        helper([2, 3, 0, 7], "mean", nan)
+        helper([2, 3, 5, 0], "mean", nan)
+
+        zero = torch.tensor([0], device=device).resize_([])
+        helper([0, 3], "sum", zero)
+        helper([0, 3, 5, 7], "sum", zero)
+        helper([2, 3, 0, 7], "sum", zero)
+        helper([2, 3, 5, 0], "sum", zero)
```
[the remark this answers is not in the document]
Comment (pbelevich) [reply test/test_nn.py]: addressed

File: test/test_nn.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply test/test_nn.py]: Everything you said is correct. I was suggesting that instead of `torch.tensor([float('nan')], device=device).resize_([])`, we use `torch.tensor(float('nan'), device=device)`, because those two expressions are equivalent (they both have size []). For example:
```
import torch
nan1 = torch.tensor([float('nan')]).resize_([])
nan2 = torch.tensor(float('nan'))

>>> print(nan1.size())
torch.Size([])

>>> print(nan2.size())
torch.Size([])
```

File: aten/src/THCUNN/ClassNLLCriterion.cu (hunk 16 above)
Comment (pbelevich) [reply aten/src/THCUNN/ClassNLLCriterion.cu]: addressed

File: test/test_nn.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (pbelevich) [reply test/test_nn.py]: thanks! addressed

File: test/test_nn.py [hunk 19]
```diff
@@ -10465,6 +10452,70 @@ def test_multi_margin_loss_errors(self, device):
...
+    def test_nll_loss_empty_tensor_reduction_mean(self, device):
+        nan = torch.tensor(float('nan'), device=device)
+        self._nll_loss_helper([0, 3], "mean", nan, device)
+        self._nll_loss_helper([0, 3, 5, 7], "mean", nan, device)
+        self._nll_loss_helper([2, 3, 0, 7], "mean", nan, device)
+        self._nll_loss_helper([2, 3, 5, 0], "mean", nan, device)
+        self._nll_loss_helper([2, 3, 5, 7, 0], "mean", nan, device)
+
+    def test_nll_loss_empty_tensor_reduction_sum(self, device):
+        zero = torch.tensor([0], device=device).resize_([])
```
Comment (zou3519) [inline test/test_nn.py]: This can be `zero = torch.tensor(0, device=device)`

File: test/test_nn.py [hunk 20]
```diff
@@ -10465,6 +10452,70 @@ def test_multi_margin_loss_errors(self, device):
...
+        self._nll_loss_helper([0, 3], "sum", zero, device)
+        self._nll_loss_helper([0, 3, 5, 7], "sum", zero, device)
+        self._nll_loss_helper([2, 3, 0, 7], "sum", zero, device)
+        self._nll_loss_helper([2, 3, 5, 0], "sum", zero, device)
+        self._nll_loss_helper([2, 3, 5, 7, 0], "sum", zero, device)
+
+    def test_nll_loss_zero_tensor(self, device):
```
Comment (zou3519) [inline test/test_nn.py]: This PR fixes nll_loss to support empty tensors on GPU. It is not clear to me what this case is testing or how it relates to nll_loss support on GPU.

File: aten/src/THCUNN/ClassNLLCriterion.cu [hunk 21]
```diff
@@ -125,7 +125,7 @@ __global__ void cunn_ClassNLLCriterion_updateOutput_kernel(Dtype *output,
     }
     *total_weight = ScalarConvert<Acctype, Dtype>::to(total_weightAcc);
     *output = ScalarConvert<Acctype, Dtype>::to(outputAcc);
-    if (size_average && *total_weight > 0) {
+    if (size_average && ((*total_weight != 0 && outputAcc != 0) || nframe == 0)) {
```
Comment (zou3519) [inline aten/src/THCUNN/ClassNLLCriterion.cu]: Why does this expression need to be this complicated?

File: aten/src/THCUNN/SpatialClassNLLCriterion.cu [hunk 22]
```diff
@@ -121,10 +121,12 @@ __global__ void cunn_SpatialClassNLLCriterion_updateOutput_kernel(
 template<typename T>
 __global__ void cunn_SpatialClassNLLCriterion_sizeAverage_kernel(
           T *output,
-          T *total_weight)
+          T *total_weight,
+          int nElement)
 {
-  if (*total_weight > 0)
+  if ((*total_weight != 0 && output != 0) || nElement == 0) {
```
Comment (zou3519) [inline aten/src/THCUNN/SpatialClassNLLCriterion.cu]: Ditto, this is getting complicated. The downside here is that it'll be difficult to maintain in the future -- why do we need to do this?

File: test/test_nn.py (hunk 20 above)
Comment (pbelevich) [reply test/test_nn.py]: It is generalized version of what you asked in [previous review](https://github.com/pytorch/pytorch/pull/31491#discussion_r365461289)

Comment (zou3519) [reply test/test_nn.py]: I see, thanks for clarifying. Can we name this something more descriptive? Something like `test_nll_loss_total_weight_is_zero`.

File: aten/src/THCUNN/ClassNLLCriterion.cu (hunk 21 above)
Comment (pbelevich) [reply aten/src/THCUNN/ClassNLLCriterion.cu]: It is what you asked me to fix in [previous review](https://github.com/pytorch/pytorch/pull/31491#discussion_r365461289)

File: aten/src/THCUNN/SpatialClassNLLCriterion.cu (hunk 22 above)
Comment (pbelevich) [reply aten/src/THCUNN/SpatialClassNLLCriterion.cu]: It is what you asked me to fix in [previous review](https://github.com/pytorch/pytorch/pull/31491#discussion_r365461289)

File: aten/src/THCUNN/ClassNLLCriterion.cu (hunk 21 above)
Comment (zou3519) [reply aten/src/THCUNN/ClassNLLCriterion.cu]: Maybe something like 
```
if (size_average) {
    if (nframe == 0) {
        *output = NaN;
    }
    if (*total_weight != 0) {
        *output = ScalarConvert<Acctype, Dtype>::to(outputAcc / total_weightAcc);
    }
}
```
makes the logic clearer

**Review by zou3519 — APPROVED** [review]
